images/executor/src/index.ts

```diff
@@ -0,0 +1,112 @@
+import readline from "readline";
+import path from "path";
+import {
+  PythonCaller,
+  CallerOptions,
+  FunctionMissingError
+} from "./code-caller";
+
+interface Request {
+  file: string;
+  fcn: string;
+  args: any[];
+  cwd: string;
+  paths: string[];
+}
+
+interface Results {
+  error?: string;
+  data?: any;
+  output?: string;
+  functionMissing?: boolean;
+  needsFullRestart: boolean;
+}
+
+async function handleInput(
+  line: string,
+  caller: PythonCaller
+): Promise<Results> {
+  return new Promise((resolve, reject) => {
+    const request: Request = JSON.parse(line);
+
+    if (request.fcn === "restart") {
+      caller.restart((restartErr, success) => {
+        resolve({
+          data: "success",
+          needsFullRestart: !!restartErr || !success
+        });
+      });
+      return;
+    }
+
+    const options: CallerOptions = {
+      paths: request.paths || [],
+      cwd: request.cwd,
+      timeout: 20000 // Probably too high; just copied from PrairieLearn
+    };
+
+    caller.call(
+      request.file,
+      request.fcn,
+      request.args,
+      options,
+      (err, data, output) => {
+        const functionMissing = err instanceof FunctionMissingError;
+        resolve({
+          // `FunctionMissingError` shouldn't be propagated as an actual error
+          // we'll report it via `functionMissing`
+          error: err && !functionMissing ? err.message : undefined,
+          data,
+          output,
+          functionMissing,
+          needsFullRestart: false
+        });
+      }
+    );
+  });
+}
+
+(async () => {
```

Comment:
Does this need to be `async`? We aren’t using `await` anywhere inside? Of course I’d prefer it we did use `await` and so this `async` was necessary 😄

Reply:
It does not! I addressed this in 717e4ec.